File: full_case_study_solns/_f_planar_vtol/python_old/hw8/VTOLController.py
import numpy as np
import sys
sys.path.append('..')  # add parent directory
import VTOLParam as P
import VTOLParamHW8 as P8
from PDControl import PDControl

class VTOLController:
    ''' 
        This class inherits other controllers in order to organize multiple controllers.
    '''

    def __init__(self):
        self.zCtrl = PDControl(P8.kp_z, P8.kd_z, P.fmax, P.beta, P.Ts)
        self.hCtrl = PDControl(P8.kp_h, P8.kd_h, P.fmax, P.beta, P.Ts)
        self.thetaCtrl = PDControl(P8.kp_th, P8.kd_th, P.fmax, P.beta, P.Ts)

    def u(self, r, y):
        z_r = float(r[0])
        h_r = float(r[1])
        z = y[0]
        h = y[1]
        theta = y[2]
        F_tilde = self.hCtrl.PD(h_r, h, flag=False)
        F = F_tilde + P.Fe
        theta_ref = self.zCtrl.PD(z_r, z, flag=False)
        tau = self.thetaCtrl.PD(theta_ref, theta, flag=False)
        return np.matrix([[F], [tau]])

File: full_case_study_solns/_f_planar_vtol/matlab_old/hw8/VTOLController.m
classdef VTOLController
    %----------------------------
    properties
        zCtrl
        hCtrl
        thetaCtrl
        mixing
        Fe
    end
    %----------------------------
    methods
        %----------------------------
        function self = VTOLController(P)
            self.Fe = P.Fe;
            self.mixing = P.mixing;
            % Instantiates the PD control objects
            self.zCtrl = PDControl(P.kp_z, P.kd_z, P.fmax, P.beta, P.Ts);
            self.hCtrl = PDControl(P.kp_h, P.kd_h, P.fmax, P.beta, P.Ts);
            self.thetaCtrl = PDControl(P.kp_th, P.kd_th, P.fmax, P.beta, P.Ts);
        end
        %----------------------------
        function out = u(self, r, y)
            z_r = r(1);
            h_r = r(2);
            z = y(1);
            h = y(2);
            theta = y(3);

            F_tilde = self.hCtrl.PD(h_r, h, false);
            theta_ref = self.zCtrl.PD(z_r, z, false);
            tau = self.thetaCtrl.PD(theta_ref, theta, false);
            
            % total force
            F = F_tilde + self.Fe;
            
            % need to saturate the individual motors.  Convert back to f_r
            % and f_l to saturate each command individually, then turn back
            % to the saturated tau, F values.
            out = [F; tau];
            u = self.mixing*out;
            u(1) = self.thetaCtrl.saturate(u(1));
            u(2) = self.thetaCtrl.saturate(u(2));
            out = inv(self.mixing)*u;
                        
        end
    end
end